### PR TITLE
[main < T518] Add algo.all_simple_paths

### PR DESCRIPTION
**Pull request by imilinovic**
### Description

Add algo all_simple_paths

### Pull request type

- [ ] Bugfix
- [X] Algorithm/Module
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Related issues

Delete if this PR doesn't resolve any issues. Link the issue if it does.

######################################

#### Module/Algorithm
- [ ] Core algorithm/module implementation
- [ ] [Query module](https://memgraph.com/docs/memgraph/reference-guide/query-modules) implementation
- [ ] Unit tests
- [ ] End-to-end tests
- [ ] Code documentation
- [ ] README short description
- [ ] Documentation on [memgraph/docs](https://github.com/memgraph/docs)
- [ ] Update GQLALchemy signatures in [query builder](https://github.com/memgraph/gqlalchemy/blob/main/gqlalchemy/graph_algorithms/query_builder.py  ) using [query module signature generator](https://github.com/memgraph/gqlalchemy/blob/main/scripts/query_module_signature_generator.py)

######################################

### CODE REVIEW

**Comment by imilinovic** [comment]
Requires:
[Memgraph PR1249](https://github.com/memgraph/memgraph/pull/1249)

File: cpp/algo_module/algorithm/algo.cpp [hunk 1]
```diff
@@ -1,5 +1,132 @@
...
+
+  iterate(curr_node.InRelationships(), false);
+  iterate(curr_node.OutRelationships(), true);
+  visited.erase(curr_node.Id().AsInt());
+}
+
+void Algo::PathFinder::FindAllPaths() {
+  mgp::Path path{start_node_};
+  std::unordered_set<int64_t> visited;
+  DFS(start_node_, path, visited);
```
Comment (antoniofilipovic) [inline cpp/algo_module/algorithm/algo.cpp:107]: Maybe BFS?

File: cpp/algo_module/algorithm/algo.cpp [hunk 2]
```diff
@@ -1,5 +1,132 @@
...
+    any_incoming_ = true;
+    return;
+  }
+
+  for (const auto &rel : relationship_types) {
+    std::string rel_type{std::string(rel.ValueString())};
+    bool starts_with = rel_type.starts_with('<');
+    bool ends_with = rel_type.ends_with('>');
+
+    if (rel_type.size() == 1) {
```
Comment (antoniofilipovic) [inline cpp/algo_module/algorithm/algo.cpp:23]: Can you comment here if `<` or `>` is sent, then for all relationships `any_incoming_` or `any_outgoing_` works

File: cpp/algo_module/algorithm/algo.cpp [hunk 3]
```diff
@@ -1,5 +1,132 @@
...
+        } else if (!outgoing) {
+          seen.insert(relationship.From().Id().AsInt());
+        }
+      }
+    }
+  };
+
+  iterate(curr_node.InRelationships(), false);
+  iterate(curr_node.OutRelationships(), true);
+  visited.erase(curr_node.Id().AsInt());
```
Comment (antoniofilipovic) [inline cpp/algo_module/algorithm/algo.cpp:101]: Try to check here if user sent at all InRelationship, and if not that we don't expand on in relationships

**Review by antoniofilipovic — CHANGES_REQUESTED** [review]
All looks good, but left two comments to make it more optimal

**Review by antoniofilipovic — APPROVED** [review]

**Pull request merged**